subsys/usb/usb_c/usbc_pe_src_states.c

```diff
@@ -0,0 +1,702 @@
+	pe_set_state(dev, PE_SRC_STARTUP);
+}
+
+/**
+ * @brief Handle source-specific DPM requests
+ */
+bool source_dpm_requests(const struct device *dev)
+{
+	struct usbc_port_data *data = dev->data;
+	struct policy_engine *pe = data->pe;
+
+	if (pe->dpm_request > REQUEST_TC_END) {
+		atomic_set_bit(pe->flags, PE_FLAGS_DPM_INITIATED_AMS);
+
+		if (pe->dpm_request == REQUEST_GET_SNK_CAPS) {
+			pe_set_state(dev, PE_GET_SINK_CAP);
+		} else if (pe->dpm_request == REQUEST_PE_GOTO_MIN) {
+			pe_set_state(dev, PE_SRC_TRANSITION_SUPPLY);
+		}
+		return true;
+	}
+
+	return false;
+}
+
+/**
+ * @brief Send Source Caps to Sink
+ */
+static void send_src_caps(struct policy_engine *pe)
+{
+	const struct device *dev = pe->dev;
+	struct usbc_port_data *data = dev->data;
+	struct protocol_layer_tx_t *prl_tx = data->prl_tx;
+	struct pd_msg *msg = &prl_tx->emsg;
+	const uint32_t *pdos;
+	uint32_t num_pdos = 0;
+
+	/* This callback must be implemented */
+	__ASSERT(data->policy_cb_get_src_caps != NULL,
+			"Callback pointer should not be NULL");
+
+	data->policy_cb_get_src_caps(dev, &pdos, &num_pdos);
+
+	msg->len = PD_CONVERT_PD_HEADER_COUNT_TO_BYTES(num_pdos);
+	memcpy(msg->data, (uint8_t *)pdos, msg->len);
+	pe_send_data_msg(dev, PD_PACKET_SOP, PD_DATA_SOURCE_CAP);
+}
+
+/**
+ * @brief 8.3.3.2.1 PE_SRC_Startup State
+ */
+void pe_src_startup_entry(void *obj)
+{
+	struct policy_engine *pe = (struct policy_engine *)obj;
+	const struct device *dev = pe->dev;
+
+	LOG_INF("PE_SRC_Startup");
+
+	/* Reset CapsCounter */
+	pe->caps_counter = 0;
+
+	/* Reset the protocol layer */
+	prl_reset(dev);
+
+	/* Set power role to Source */
+	pe->power_role = TC_ROLE_SOURCE;
+
+	/* Invalidate explicit contract */
+	atomic_clear_bit(pe->flags, PE_FLAGS_EXPLICIT_CONTRACT);
+
+	policy_notify(dev, NOT_PD_CONNECTED);
+}
+
+void pe_src_startup_run(void *obj)
+{
+	struct policy_engine *pe = (struct policy_engine *)obj;
+	const struct device *dev = pe->dev;
+
+	/*
+	 * Once the reset process completes, the Policy Engine Shall
+	 * transition to the PE_SRC_Send_Capabilities state
+	 */
+	if (prl_is_running(dev)) {
+		pe_set_state(dev, PE_SRC_SEND_CAPABILITIES);
+	}
+}
+
+/**
+ * @brief 8.3.3.2.2 PE_SRC_Discover State
+ */
+void pe_src_discovery_entry(void *obj)
+{
+	struct policy_engine *pe = (struct policy_engine *)obj;
+
+	LOG_INF("PE_SRC_Discover");
+
+	/*
+	 * Start the SourceCapabilityTimer in order to trigger sending a
+	 * Source_Capabilities message
+	 */
+	usbc_timer_start(&pe->pd_t_typec_send_source_cap);
+}
+
+void pe_src_discovery_run(void *obj)
+{
+	struct policy_engine *pe = (struct policy_engine *)obj;
+	const struct device *dev = pe->dev;
+
+	/*
+	 * The Policy Engine Shall transition to the PE_SRC_Send_Capabilities state when:
+	 *	1) The SourceCapabilityTimer times out
+	 *	2) And CapsCounter ≤ nCapsCount
+	 */
+	if (usbc_timer_expired(&pe->pd_t_typec_send_source_cap)
+			&& pe->caps_counter <= PD_N_CAPS_COUNT) {
+		pe_set_state(dev, PE_SRC_SEND_CAPABILITIES);
+	}
```

Comment:
entering the disabled state when nCapsCount is reached is optional behavior. The nHardResetCount his handled in the super state.

Reply:
Ack. I left open the comment in the super state about how this could be easier to follow with different organization.

Do you plan to add a transition to PE_SRC_Disabled in a later PR or leave it as is?